### PR TITLE
Do not treat bare multisig outputs as IsMine unless watched

### PR DESCRIPTION
**Pull request by sipa**
Currently our wallet code will treat bare multisig outputs (meaning scriptPubKeys with multiple public keys + `OP_CHECKMULTISIG` operator in it) as ours without the user asking for it, as long as all private keys in it are in our wallet.

This is a pointless feature. As it only works when all private keys are in one place, it's useless compared to single key outputs (P2PK, P2PKH, P2WPKH, P2SH-P2WPKH), and worse in terms of space, cost, UTXO size, and ability to test (due to lack of address format for them).

Furthermore, they are problematic in that producing a list of all `scriptPubKeys` we accept is not tractable (it involves all combinations of all public keys that are ours). In further wallet changes I'd like to move to a model where all scriptPubKeys that are treated as ours are explicit, rather than defined by whatever keys we have. The current behavior of the wallet is very hard to model in such a design, so I'd like to get rid of it.

I think there are two options:
* Remove it entirely (do not ever accept bare multisig outputs as ours, unless watched)
* Only accept bare multisig outputs in situations where the P2SH version of that output would also be acceptable

This PR implements the first option. The second option was explored in #12874.

### CODE REVIEW

File: src/script/ismine.cpp [hunk 1]
```diff
@@ -22,25 +29,7 @@ static bool HaveKeys(const std::vector<valtype>& pubkeys, const CKeyStore& keyst
...
-}
-
-isminetype IsMine(const CKeyStore &keystore, const CTxDestination& dest, bool& isInvalid, SigVersion sigversion)
-{
-    CScript script = GetScriptForDestination(dest);
-    return IsMine(keystore, script, isInvalid, sigversion);
-}
-
-isminetype IsMine(const CKeyStore &keystore, const CScript& scriptPubKey, bool& isInvalid, SigVersion sigversion)
+static isminetype IsMineInner(const CKeyStore &keystore, const CScript& scriptPubKey, bool& isInvalid, IsMineSigVersion sigversion)
```
Comment (promag) [inline src/script/ismine.cpp]: `const CKeyStore& keystore`

File: src/script/ismine.cpp [hunk 2]
```diff
@@ -150,3 +142,26 @@ isminetype IsMine(const CKeyStore &keystore, const CScript& scriptPubKey, bool&
     }
     return ISMINE_NO;
 }
+
+isminetype IsMine(const CKeyStore &keystore, const CScript& scriptPubKey, bool& isInvalid)
```
Comment (promag) [inline src/script/ismine.cpp]: `const CKeyStore& keystore`

File: src/script/ismine.cpp [hunk 3]
```diff
@@ -150,3 +142,26 @@ isminetype IsMine(const CKeyStore &keystore, const CScript& scriptPubKey, bool&
...
+    return IsMineInner(keystore, scriptPubKey, isInvalid, IsMineSigVersion::BASE);
+}
+
+isminetype IsMine(const CKeyStore& keystore, const CScript& scriptPubKey)
+{
+    bool isInvalid = false;
+    return IsMine(keystore, scriptPubKey, isInvalid);
+}
+
+isminetype IsMine(const CKeyStore &keystore, const CTxDestination& dest, bool& isInvalid)
```
Comment (promag) [inline src/script/ismine.cpp]: `const CKeyStore& keystore`

**Review by promag — COMMENTED** [review]
Some nits…

**Comment by sipa** [comment]
Addressed @promag's nits, and added a few more code cleanups.

File: doc/release-notes.md [hunk 4]
```diff
@@ -96,6 +96,15 @@ Low-level RPC changes
...
   name of any wallet is just its `<path>` string.
 
+- Bare multisig outputs to our keys are no longer automatically treated as
+  incoming payments. As this feature was only available for multisig outputs for
+  which you had all private keys in your wallet, there was generally no use for
+  them compared to single-key schemes. Furthermore, no address format for such
+  outputs is defined, and wallet software can't easily send to it. These outputs
+  will no longer show up in `listtransactions`, `listunspent`, or contribute to
+  your balance, unless they are explicitly watched (using `importaddress` with
+  hex script argument). `signrawtransaction*` also still works for them.
```
Comment (promag) [inline doc/release-notes.md]: `` `importaddress` or `importmulti`?``

**Review by promag — COMMENTED** [review]
utACK 7e827e4.

Change looks good. Commit 000c41a is nice because it also prevents repetitive hash computation. One nit fix is in the wrong commit though.

**Comment by sipa** [comment]
Addressed @promag's nits.

**Comment by promag** [comment]
@sipa thanks.

Code wise ACK 2687bd0.

File: src/script/ismine.cpp [hunk 5]
```diff
@@ -70,7 +58,7 @@ isminetype IsMine(const CKeyStore &keystore, const CScript& scriptPubKey, bool&
             // This also applies to the P2WSH case.
             break;
         }
-        isminetype ret = ::IsMine(keystore, GetScriptForDestination(CKeyID(uint160(vSolutions[0]))), isInvalid, SigVersion::WITNESS_V0);
+        isminetype ret = ::IsMineInner(keystore, GetScriptForDestination(CKeyID(uint160(vSolutions[0]))), isInvalid, SigVersion::WITNESS_V0);
```
Comment (jimpo) [inline src/script/ismine.cpp]: commit: Do not expose SigVersion argument to IsMine

nit: Drop the `::`. It's weird to have it on this call but not the other two.

File: src/script/ismine.cpp [hunk 6]
```diff
@@ -13,6 +13,12 @@
 
 typedef std::vector<unsigned char> valtype;
 
+enum class IsMineSigVersion
```
Comment (jimpo) [inline src/script/ismine.cpp]: commit: Switch to a private version of SigVersion inside IsMine

Please let a comment explaining what this is and how it differs from SigVersion. Even though it's in the commit comment, it's good to have it in the code.

File: src/script/ismine.cpp [hunk 7]
```diff
@@ -49,7 +50,7 @@ static isminetype IsMineInner(const CKeyStore& keystore, const CScript& scriptPu
         break;
     case TX_PUBKEY:
         keyID = CPubKey(vSolutions[0]).GetID();
-        if (sigversion != IsMineSigVersion::BASE && vSolutions[0].size() != 33) {
+        if (sigversion != IsMineSigVersion::TOP && sigversion != IsMineSigVersion::P2SH && vSolutions[0].size() != 33) {
```
Comment (jimpo) [inline src/script/ismine.cpp]: commit: Track difference between scriptPubKey and P2SH execution in IsMine

I think a function `bool SigVersionPermitsUncompressed(IsMineSigVersion)` would improve clarity.

**Review by jimpo — COMMENTED** [review]
Looks good.

File: src/script/ismine.cpp (hunk 5 above)
Comment (sipa) [reply src/script/ismine.cpp]: Fixed.

File: src/script/ismine.cpp (hunk 6 above)
Comment (sipa) [reply src/script/ismine.cpp]: Done (in the next commit, which actually introduces the difference).

File: src/script/ismine.cpp (hunk 7 above)
Comment (sipa) [reply src/script/ismine.cpp]: Yes, that's much more readable. Done.

**Comment by jimpo** [comment]
utACK 7d0f80b

**Comment by promag** [comment]
utACK 7d0f80b.

**Comment by laanwj** [comment]
utACK 3deed79806bbf31e003444798d9e70efffe7b8f7

**Pull request merged**